### PR TITLE
Add type stubs for multiprocessing.Manager()

### PR DESCRIPTION
**Pull request by jreese**

### CODE REVIEW

**Comment by jreese** [comment]
This currently fails the mypy test with the following output: https://docs.python.org/3/library/multiprocessing.html#multiprocessing-managers

However, those errors seems like a bug in typeshed/mypy, caused by class methods named identically to the class types that they return.  I saw similar issues when doing `from queue import Queue` and `def Queue(self) -> Queue: ...`.  Is this expected behavior, or is there a known workaround for this?

**Comment by gvanrossum** [comment]
So for dict and list you can use Dict and List. For Namespace you may have to resort to defining a helper alias _Nameapace.

**Comment by jreese** [comment]
Updated types, added `_Namespace` alias, and tests are passing.

File: stdlib/3/multiprocessing/managers.pyi [hunk 1]
```diff
@@ -2,9 +2,29 @@
...
 class BaseManager:
     def register(self, typeid: str, callable: Any = ...) -> None: ...
 
+class SyncManager(BaseManager):
+    def BoundedSemaphore(self, value: Any = ...) -> threading.BoundedSemaphore: ...
+    def Condition(self, lock: Any = ...) -> threading.Condition: ...
+    def Event(self) -> threading.Event: ...
+    def Lock(self) -> threading.Lock: ...
+    def Namespace(self) -> _Namespace: ...
+    def Queue(self, maxsize: int = 0) -> queue.Queue: ...
```
Comment (JelleZijlstra) [inline stdlib/3/multiprocessing/managers.pyi]: Defaults should always be `...`.

File: stdlib/3/multiprocessing/managers.pyi [hunk 2]
```diff
@@ -2,9 +2,29 @@
...
+    def RLock(self) -> threading.RLock: ...
+    def Semaphore(self, value: Any = ...) -> threading.Semaphore: ...
+    def Array(self, typecode: Any, sequence: Any) -> Any: ...
+    def Value(self, typecode: Any, value: Any) -> Any: ...
+    def dict(self, sequence: Any = ...) -> Dict: ...
```
Comment (JelleZijlstra) [inline stdlib/3/multiprocessing/managers.pyi]: Can we be more specific than `Dict` here? What are the keys and values? Same for List below.

Comment (jreese) [inline stdlib/3/multiprocessing/managers.pyi]: I'm not sure I understand; do you mean something like `Dict[Any, Any]` or `List[Any]`?

Comment (JelleZijlstra) [inline stdlib/3/multiprocessing/managers.pyi]: Yes, preferably with something other than `Any` if that's possible.

**Comment by jreese** [comment]
Updated with use of TypeVar definitions for mapping/sequence types

File: stdlib/3/multiprocessing/managers.pyi [hunk 3]
```diff
@@ -2,9 +2,33 @@
...
     def register(self, typeid: str, callable: Any = ...) -> None: ...
 
+class SyncManager(BaseManager):
+    def BoundedSemaphore(self, value: Any = ...) -> threading.BoundedSemaphore: ...
+    def Condition(self, lock: Any = ...) -> threading.Condition: ...
+    def Event(self) -> threading.Event: ...
+    def Lock(self) -> threading.Lock: ...
+    def Namespace(self) -> _Namespace: ...
+    def Queue(self, maxsize: int = ...) -> queue.Queue: ...
+    def RLock(self) -> threading.RLock: ...
```
Comment (JelleZijlstra) [inline stdlib/3/multiprocessing/managers.pyi]: This actually isn't exactly right, it returns an `AcquirerProxy` of some sort. The other methods here also return proxies of some sort, and that's documented at https://docs.python.org/3/library/multiprocessing.html#multiprocessing.managers.SyncManager.

It's documented that these proxies mostly behave like the objects they proxy, so I think it's probably fine to just pretend they return the objects directly.

**Comment by jreese** [comment]
Rebased on latest master to resolve  merge conflicts.  Are there any further changes that I still need to make for this to land?

**Comment by JelleZijlstra** [comment]
Thanks! I'll merge once Travis passes.

**Pull request merged**